tests/integration_test.go

```diff
@@ -0,0 +1,386 @@
+package tests
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+
+	"github.com/mikewurtz/taskman/certs"
+	pb "github.com/mikewurtz/taskman/gen/proto"
+	"github.com/mikewurtz/taskman/internal/grpc/server"
+)
+
+const (
+	testUserID = "client001"
+)
+
+var (
+	stopServer     func()
+	testServerAddr string
+	once           sync.Once
+)
+
+func TestMain(m *testing.M) {
+	code := m.Run()
+
+	if stopServer != nil {
+		stopServer()
+	}
+	os.Exit(code)
+}
+
+// startTestServer starts the test server and returns a function to stop it
+// will only be called once
+func startTestServer(t *testing.T) {
+	once.Do(func() {
+		fmt.Println("starting test server")
+		srv, err := server.NewServer("localhost:0")
+		require.NoError(t, err, "failed to create test server")
+
+		go func() {
+			if err := srv.Start(); err != nil {
+				fmt.Printf("Test server error: %v\n", err)
+			}
+		}()
+
+		testServerAddr = srv.Addr()
+		// wait for the server to be ready to handle requests, 2 seconds should be plenty
+		require.Eventually(t, func() bool {
+			client := createTestClient(t, testUserID)
+			// set context timeout to 500ms so we can still retry a few times if it fails
+			ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+			defer cancel()
+
+			// try a get task status, if it returns codes.Unimplemented then the server is ready
+			_, err := client.GetTaskStatus(ctx, &pb.TaskStatusRequest{
+				TaskId: "375b0522-72ed-4f3f-88d0-01d360d06b8c",
+			})
+
+			sts, ok := status.FromError(err)
+			// TODO: this will need to be updated once we have a real implementation
+			return ok && sts.Code() == codes.Unimplemented
+		}, 2*time.Second, 10*time.Millisecond, "gRPC server did not start in time")
+
+		stopServer = func() {
```

Comment:
I can see why you used the sync.Once to share a single server between all tests, but I'm not sure the extra complexity is worth it here.

One thing I don't love about this approach is it creates asymmetry between `startTestServer` (which is called by each test), and `stopServer` (which is called once).

Reply:
I agree with, I can remove the sync.Once and move the startTestServer bringup into the TestMain function. That way it happens once just like stopServer. I will probably need to have my own retry loop instead of `require.Eventually` since it requires an actual t.TestingT object and I do not believe that is accessible in TestMain.